File: 0038-count-and-say/0038-count-and-say.js
/**
 * @param {number} n
 * @return {string}
 */
var countAndSay = function(n) {
    
    let say = (str) => {
        //console.log(str);
        let stack = [];
        for(let ch of str) {
            if(stack.length <= 0) {
                stack.push([ch,1]);
                continue;
            } else {
                let top = stack[stack.length-1];
                if(ch == top[0]) top[1]++;
                else stack.push([ch,1]);
            }
        }
        
        //console.log(stack);
        let ans = [];
        for(let el of stack) {
            ans.push(el[1]);
            ans.push(el[0]);
        }
        return ans.join('');
    }
    let dp = (i) => {
        if( i == 1) {
            return "1";
        }
        
        return say(dp(i-1));
    }
    return dp(n);
};

File: 0038-count-and-say/0038-count-and-say.java
class Node {
    public int count;
    public char ch;
    public Node(int count, char ch) {
        this.count = count;
        this.ch = ch;
    }
}

class Solution {
    public String say(String s) {
        Stack<Node> stack = new Stack<Node>();
        for(int i = 0; i < s.length(); i++) {
            if(stack.isEmpty()) {
                stack.push(new Node(1, s.charAt(i)));
                continue;
            } else {
                Node top = stack.peek();
                if(s.charAt(i) == top.ch) {
                    top.count++;
                } else {
                    stack.push(new Node(1, s.charAt(i)));
                }
            }
        }
        StringBuffer bf = new StringBuffer();
        for(Node el : stack) {
            bf.append(el.count + "" + el.ch);
        }
        return bf.toString();
    }
    public String dp(int n) {
        if(n == 1) {
            return "1";
        }
        return say(dp(n-1));
    }
    public String countAndSay(int n) {
        return dp(n);
    }
}